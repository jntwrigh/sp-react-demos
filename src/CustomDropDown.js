import React, {Component} from 'react';

class CustomDropDown extends Component {
  constructor( props ) {
    super( props );

    this.idField = props.idField || 'id';

    let defaultValue = props.defaultValue || props.value || '';
    this.placeholder = props.placeholder || 'Please select an item below';

    this.state = {
      value: defaultValue,
    };

    this.domProps = {
      className: props.className,
      name     : props.name,
      onChange : props.onChange,
    };

    this.handleChange = this.handleChange.bind( this );
  }

  getLabel( option, label ) {
    let labels = label.split( /\s+/ );
    return labels.map( label => option[ label ] ).join( ' ' );
  }

  handleChange( event ) {
    this.setState( { value: event.target.value } );
    if ( this.props.onChange ) this.props.onChange( event );
  }

  render() {
    return (
      <select {...this.domProps} onChange={this.handleChange} value={this.state.value}>
        <option value="">{this.placeholder}</option>
        {this.props.options.map( ( option, index ) => (
          <option value={option[ this.idField ]} key={index}>{this.getLabel( option, this.props.label )}</option>
        ) )}
      </select>
    );
  }
}

export default CustomDropDown;
