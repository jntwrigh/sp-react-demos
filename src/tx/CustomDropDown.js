import React, {Component} from 'react';

class CustomDropDown extends Component {
  constructor( props ) {
    super( props );

    this.idField = props.idField || 'id';

    this.handleChange = this.handleChange.bind(this);
  }

  getLabel( option, label ) {
    let labels = label.split( /\s+/ );
    return labels.map( label => option[ label ] ).join( ' ' );
  }

  handleChange(event) {
    let found = this.props.options.find(option => option[this.idField] === Number(event.target.value));

    console.log('CustomDropDown selected: ', found);
  }

  render() {
    return (
      <select onChange={this.handleChange}>
        {this.props.options.map( (option, index) => (
          <option value={option[ this.idField ]} key={index}>{this.getLabel( option, this.props.label )}</option>
        ) )}
      </select>
    );
  }
}

export default CustomDropDown;
