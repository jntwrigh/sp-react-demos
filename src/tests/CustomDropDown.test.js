import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import CustomDropDown from '../CustomDropDown';
import { peopleDAO } from '../data/class-data';

import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({adapter: new Adapter()});

let personList;

beforeEach( () => {
  personList = peopleDAO.list();
} );

test( 'CustomDropDown single labels render correctly', () => {
  const dropDown = shallow(
    <CustomDropDown idField="id" label="lastName" options={ personList }/>
  );

  expect( dropDown.text() ).toMatch( new RegExp( personList[ 0 ].lastName ) );

  // Negative confirmation
  expect( dropDown.text() ).not.toMatch( /Paxton/ );
} );

test( 'CustomDropDown multiple labels render correctly', () => {
  const dropDown = shallow(
    <CustomDropDown idField="id" label="firstName lastName" options={ personList }/>
  );

  expect( dropDown.text() ).toMatch( new RegExp( `${personList[ 0 ].firstName} ${personList[ 0 ].lastName}` ) );

  // Negative confirmation
  expect( dropDown.text() ).not.toMatch( /Paxton/ );
} );

it( 'CustomDropDown selections work', () => {
  const dropDown = shallow(
    <CustomDropDown idField="id" label="lastName" options={ personList }/>
  );

  const select = dropDown.find( 'select' );
  select.simulate( 'change', { target: { value: '203' } } );

  expect( dropDown.state( 'value' ) ).toEqual( '203' );

} );

it( 'CustomDropDown defaultValue test', () => {
  const dropDown = shallow(
    <CustomDropDown idField="id" label="lastName" options={ personList } defaultValue="203"/>
  );

  expect( dropDown.state( 'value' ) ).toEqual( '203' );
} );

// Snapshot test
it( 'CustomDropDown snapshot test', () => {
  const tree = renderer
    .create( <CustomDropDown idField="id" label="lastName" options={ personList } defaultValue="203"/> )
    .toJSON();

  expect( tree ).toMatchSnapshot();

} );

it( 'CustomDropDown placeholder snapshot', () => {
  const tree = renderer
    .create( <CustomDropDown idField="id"
                             label="lastName"
                             options={ personList }
                             defaultValue="203"
                             placeholder="Select a person"/> )
    .toJSON();

  expect( tree ).toMatchSnapshot();

} );
