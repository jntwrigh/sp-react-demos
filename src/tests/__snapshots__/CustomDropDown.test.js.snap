// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDropDown placeholder snapshot 1`] = `
<select
  className={undefined}
  name={undefined}
  onChange={[Function]}
  value="203"
>
  <option
    value=""
  >
    Select a person
  </option>
  <option
    value="201"
  >
    Sparks
  </option>
  <option
    value="202"
  >
    Trent
  </option>
  <option
    value="203"
  >
    Hawksmoor
  </option>
  <option
    value="204"
  >
    Thornedike
  </option>
  <option
    value="205"
  >
    Cortez
  </option>
  <option
    value="206"
  >
    Li-Min
  </option>
  <option
    value="207"
  >
    Jerusalem
  </option>
  <option
    value="208"
  >
    Wagner
  </option>
  <option
    value="209"
  >
    Snow
  </option>
  <option
    value="210"
  >
    Chase
  </option>
</select>
`;

exports[`CustomDropDown snapshot test 1`] = `
<select
  className={undefined}
  name={undefined}
  onChange={[Function]}
  value="203"
>
  <option
    value=""
  >
    Please select an item below
  </option>
  <option
    value="201"
  >
    Sparks
  </option>
  <option
    value="202"
  >
    Trent
  </option>
  <option
    value="203"
  >
    Hawksmoor
  </option>
  <option
    value="204"
  >
    Thornedike
  </option>
  <option
    value="205"
  >
    Cortez
  </option>
  <option
    value="206"
  >
    Li-Min
  </option>
  <option
    value="207"
  >
    Jerusalem
  </option>
  <option
    value="208"
  >
    Wagner
  </option>
  <option
    value="209"
  >
    Snow
  </option>
  <option
    value="210"
  >
    Chase
  </option>
</select>
`;
